jurity/recommenders/interlist_diversity.py

```diff
@@ -0,0 +1,214 @@
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        Data frame with (user_id, item_id) in each row.
+    user_id_column: str
+        User id column name.
+    item_id_column: str
+        Item id column name.
+
+    Returns
+    -------
+    Sparse matrix with user-item interactions/recommendations.
+    """
+
+    # Map each user_id to (0, n_users)
+    users = list(df[user_id_column].unique())
+    user_map = dict(zip(users, range(len(users))))
+
+    # Map each user_id to (0, n_items)
+    items = list(df[item_id_column].unique())
+    item_map = dict(zip(items, range(len(items))))
+
+    # Update user_id, item_id values
+    user_ids = df[user_id_column].map(user_map).values
+    item_ids = df[item_id_column].map(item_map).values
+
+    # Convert to csr matrix
+    csr_matrix = sp.coo_matrix((np.ones(len(user_ids)), (user_ids, item_ids))).tocsr()
+
+    return csr_matrix
+
+
+def interlist_diversity(predicted_results: pd.DataFrame, click_column: str, k: int,
+                        user_id_column: str = Constants.user_id, item_id_column: str = Constants.item_id,
+                        sample_size: float = None, seed: int = Constants.default_seed,
+                        chunk_size: int = 1000) -> Tuple[float, int]:
+    """
+        Calculate inter-list diversity metric:
+
+            Inter-list-diversity = 1 - average(cosine_similarity(R_{u_i}, R_{u_j})),
+            where R_{u_i} is the binary indicator vector representing provided recommendations for user u_i, i \ne j.
+
+        Parameters
+        ----------
+        predicted_results: pd.DataFrame
+            Recommendations data frame with (user_id, item_id, score) in each row.
+        k: int
+            Top-k recommendations to consider.
+        user_id_column: str
+            User id column name.
+        item_id_column: str
+            Item id column name.
+        click_column: str
+            Recommendation score column name.
+        sample_size: float
+            Proportion of users to randomly sample for evaluation.
+        seed : int
+            The seed used to create random state.
+        chunk_size: int
+            Chunk size to limit memory usage.
+
+        Returns
+        -------
+        Inter-list diversity metric, number of unique users as the support to get the metric
+        """
+
+    # Sample users
+    if sample_size is not None:
+        df = sample_users(predicted_results, user_id_column, sample_size, seed)
+    else:
+        df = predicted_results
+
+    # Sort by user and score
+    df = df.sort_values(click_column, ascending=False).groupby(user_id_column).head(k)
+
+    # Given user/item id maps, create sparse matrix.
+    sparse_matrix = tocsr(df, user_id_column, item_id_column)
+
+    # Get pairwise cosine similarities
+    similarities_sum = 0
+    if chunk_size == 1:
+        num_chunks = sparse_matrix.shape[0]
+    else:
+        num_chunks = sparse_matrix.shape[0] // chunk_size + 1
+    for i in range(num_chunks):
```

Comment:
Ah, interesting. Have you done testing on the chunk size to see if 1000 is a good default? I think this could be its own function (eg. `chunked_cosine_similarity`) and have some tests to ensure validity.

Reply:
Also saw [this](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances_chunked.html) built-in to sklearn, though I don't know if it directly satisfies our usecase.

Reply:
that sklearn function looks interesting!

Reply:
I think the sklearn function would be helpful. I would look into this. (1). it seems to work on the pairwise distances with `working_memory`-sized chunks, so we might not need to set `chunk_size` manually. When `working_memory=None`, the default memory `sklearn.get_config()['working_memory']` will be used, which is 1G by default. (2) computations can run in parallel for a speed-up. I notice a up to ~20% shorter runtime with a quick test, and using a lower version of sklearn might impact the performance I think. (3) since it calculates distances directly, the math part could be a little simplified.

Reply:
Updated the implementation with the suggested sklearn build-in function